### PR TITLE
List's traverse method was applying effects in reverse.

### LINKED ISSUE #140: is `sequence` applying effects in backwards order?
Compare:
```
val initialId = 0
case class Person(id: Int, name: String)
def allocatePersonS(name: String) = State { id: Int => (id + 1, Person(id, name)) }

scala>     val allocateThreePeople = for {
     |       p0 <- allocatePersonS("Alice")
     |       p1 <- allocatePersonS("Bob")
     |       p2 <- allocatePersonS("Charlie")
     |     } yield (p0, p1, p2)
allocateThreePeople: cats.data.State[Int,(Person, Person, Person)] = cats.data.State@24407ee8

scala>     val (id3, (p0, p1, p2)) = allocateThreePeople(initialId)
id3: Int = 3
p0: Person = Person(0,Alice)
p1: Person = Person(1,Bob)
p2: Person = Person(2,Charlie)
```
to:
```
val (id3, List(p0, p1, p2)) =
  List("Alice","Bob","Charlie")
    .map(allocatePersonS)
    .sequence[State[Int,?],Person]
    .apply(initialId)

// Exiting paste mode, now interpreting.

id3: Int = 3
p0: Person = Person(2,Alice)
p1: Person = Person(1,Bob)
p2: Person = Person(0,Charlie)
```

I haven't had a chance to check whether it's a problem with `sequence` or with `ListInstances.listInstance` or if it's something wrong that I've done; confirmation would be appreciated :)

### PR DESCRIPTION
**Pull request by non**
This commit fixes this issue (which had to do with passing
arguments to G.map2 in the wrong order) and adds a regression
test to be sure we don't accidentally mess this up in the
future.

Fixes #140 

Review by @refried et al.

### CODE REVIEW

**Comment by mpilquist** [comment]
:+1:

File: tests/src/test/scala/cats/tests/RegressionTests.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import cats._
+import cats.implicits._
+import org.scalatest.FunSuite
+
+class RegressionTests extends FunSuite {
+
+  // test sequence order using a contrived, unsafe state monad.
+  // 
+
+  test("#140: confirm sequence order") {
```
Comment (julien-truffaut) [inline tests/src/test/scala/cats/tests/RegressionTests.scala]: could make the same test on `List[Option[Int]]`?

Comment (non) [inline tests/src/test/scala/cats/tests/RegressionTests.scala]: yeah it should be possible.

**Comment by ceedubs** [comment]
I think this implementation of traverse is problematic when side effects are in play and should be executed in order. This is a problem even if the side effects are properly housed within something like `IO` or `Task`.

Example with scalaz's Task:
```scala
scala> traverse(List(1, 2, 3))(i => Task{println(i); i})
res2: scalaz.concurrent.Task[List[Int]] = scalaz.concurrent.Task@3b4c8039

scala> res2.run
3
2
1
res3: List[Int] = List(1, 2, 3)
```

I don't think you can rely on running side effects in a `map` call in the general case.

**Comment by non** [comment]
I don't think this is about side-effects, but just about the order in which the monad instances (which constrain effects) are composed, right?

**Comment by wedens** [comment]
https://github.com/scalaz/scalaz/issues/886 discussion in scalaz

**Comment by non** [comment]
The reason Scalaz hits this problem is because they need to reverse the list to do a `foldLeft`. Since we are building up a result in `ListBuffer` we can build a result directly with `foldLeft` so we don't have the same issue.

That said, I agree that side-effecting during the fold is a bad idea.

**Comment by ceedubs** [comment]
@wedens I would consider my example different than the one you have linked because I'm not using traverse directly for side effects. I'm traversing into `Task` (so basically `IO`), which as far as I know is the standard way to sequence a list of side-effecting evaluations.

@non sorry I'm not sure I completely understand your response. Are you agreeing that this is an issue, or does it not seem like a problem to you?

I have no qualms about local side effects for performance reasons. I just am not sure that you can do side effects (without using `IO` or something like it) in an arbitrary `Applicative`'s `map2` method in a reliable way.

**Pull request merged**